alphadia/fdrx/models/two_step_classifier.py

```diff
@@ -83,52 +90,94 @@ def fit_predict(
+            logger.info(f"Starting iteration {i + 1} / {self._max_iterations}.")
+
+            # extract preselction using first classifier if it is fitted
             if self.first_classifier.fitted and i > 0:
-                df_train, df_predict = self._apply_filtering_with_first_classifier(
+                df_train = self._apply_filtering_with_first_classifier(
                     df, x_cols, group_columns
                 )
+                df_predict = df_train  # using the same df for training and predicting, unlike in the following else block.
+
+                previous_target_count_after_first_clf = get_target_count(df_train)
                 self.second_classifier.epochs = 50
             else:
+                logger.info("First classifier not fitted yet. Proceeding without it.")
                 df_train = df[df["rank"] < self._train_on_top_n]
                 df_predict = df
+
                 self.second_classifier.epochs = 10
 
-            predictions = self._train_and_apply_second_classifier(
+            # train and apply second classifier
+            df_after_second_clf = self._train_and_apply_second_classifier(
                 df_train, df_predict, x_cols, y_col, group_columns
             )
 
-            # Filter results and check for improvement
-            df_filtered = filter_by_qval(predictions, self.second_fdr_cutoff)
-            current_target_count = len(df_filtered[df_filtered["decoy"] == 0])
+            df_filtered = filter_by_qval(df_after_second_clf, self.second_fdr_cutoff)
+            current_target_count = get_target_count(df_filtered)
 
-            if current_target_count < best_precursor_count:
+            if current_target_count < previous_target_count_after_second_clf:
                 logger.info(
-                    f"Stopping training after iteration {i}, "
-                    f"due to decreased target count ({current_target_count} < {best_precursor_count})"
+                    f"Training stopped on iteration {i + 1}. Decrease in target count from "
+                    f"{previous_target_count_after_second_clf:,} to {current_target_count:,}."
                 )
                 return best_result
 
-            best_precursor_count = current_target_count
-            best_result = predictions
+            previous_target_count_after_second_clf = current_target_count
+            best_result = df_after_second_clf
+
+            logger.info(
+                f"{current_target_count:,} precursors found after second classifier, "
+                f"at fdr={self.second_fdr_cutoff}"
+            )
 
-            # Update first classifier if enough confident predictions
+            # update first classifier if enough confident predictions
             if current_target_count > self._min_precursors_for_update:
-                self._update_first_classifier(
-                    df_filtered, df, x_cols, y_col, group_columns
+                logger.info(
```

Comment:
I think we can skipp this statement and only log if it wasnt the case. This will usually be true.